File: Cpp/Example3-3-WebMercatorConvertor/Example3-3-WebMercatorConvertor.cpp
﻿// Example3WebMercatorConvertor.cpp : 此文件包含 "main"
// 函数。程序执行将在此处开始并结束。
//

#include <ogr_spatialref.h>

#include <iostream>

OGRSpatialReference gcs;  //地理坐标系
OGRSpatialReference pcs;  //投影坐标系

void CreateSrs() {
  // CGCS2000
  gcs.importFromEPSG(4326);

  // Tm投影
  pcs.importFromEPSG(3857);

  //# GDAL 3 changes axis order : https://github.com/OSGeo/gdal/issues/1546
  gcs.SetAxisMappingStrategy(OAMS_TRADITIONAL_GIS_ORDER);
  pcs.SetAxisMappingStrategy(OAMS_TRADITIONAL_GIS_ORDER);
}

int main() {
  //设置Proj数据
  std::string projDataPath = getenv("GISBasic");
  projDataPath += "/share/proj";
  CPLSetConfigOption("PROJ_LIB", projDataPath.c_str());

  CreateSrs();

  OGRCoordinateTransformation* lonLat2XY =
      OGRCreateCoordinateTransformation(&gcs, &pcs);
  OGRCoordinateTransformation* xy2LonLat =
      OGRCreateCoordinateTransformation(&pcs, &gcs);
  if (!lonLat2XY || !xy2LonLat) {
    return 1;
  }

  double x = 113.6;
  double y = 38.8;
  printf("经纬度坐标：%.9lf\t%.9lf\n", x, y);
  if (!lonLat2XY->Transform(1, &x, &y)) {
    return 1;
  }
  printf("平面坐标：%.9lf\t%.9lf\n", x, y);

  if (!xy2LonLat->Transform(1, &x, &y)) {
    return 1;
  }
  printf("再次转换回的经纬度坐标：%.9lf\t%.9lf\n", x, y);

  OGRCoordinateTransformation::DestroyCT(lonLat2XY);
  lonLat2XY = nullptr;
  OGRCoordinateTransformation::DestroyCT(xy2LonLat);
  xy2LonLat = nullptr;
}

File: Python/example-3-3-web-mercator-convertor.py
import os
from numpy import double
from osgeo import osr

# 自动设置 PROJ_LIB for Conda (Windows/Linux/macOS) 
def setup_proj_lib():
    if 'CONDA_PREFIX' in os.environ:
        conda_prefix = os.environ['CONDA_PREFIX']
        # Windows
        proj_path = os.path.join(conda_prefix, 'Library', 'share', 'proj')
        if not os.path.exists(proj_path):
            # Unix-like
            proj_path = os.path.join(conda_prefix, 'share', 'proj')
        if os.path.exists(proj_path):       
            os.environ['PROJ_LIB'] = proj_path
            return True
    return False

gcs = osr.SpatialReference()
pcs = osr.SpatialReference()

def create_srs():
    # WGS84  
    gcs.ImportFromEPSG(4326)

    # Web墨卡托
    pcs.ImportFromEPSG(3857) 

    # GDAL 3 changes axis order : https://github.com/OSGeo/gdal/issues/1546
    gcs.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)
    pcs.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)

def main():
    osr.UseExceptions() # 启用异常处理（推荐），避免静默错误
    setup_proj_lib() #设置 PROJ_LIB 

    create_srs()

    lonLat2XY = osr.CoordinateTransformation(gcs, pcs)
    xy2LonLat = osr.CoordinateTransformation(pcs, gcs)
    if lonLat2XY is None or xy2LonLat is None:
        print("创建坐标转换失败")
        return 
  
    x = 113.6
    y = 38.8   
    z = 0.0
    print(f"经纬度坐标：{x:.9f}\t{y:.9f}")
    x, y, z = lonLat2XY.TransformPoint(x, y, z)     
    print(f"平面坐标：{x:.9f}\t{y:.9f}")

    x, y, z = xy2LonLat.TransformPoint(x, y, z)   
    print(f"再次转换回的经纬度坐标：{x:.9f}\t{y:.9f}")

if __name__ == "__main__":
    main()    